### PR TITLE
Introduce MzContainerized, use in reachability logging

### PR DESCRIPTION
**Pull request by antiguru**
Introduce a `MzContainerized` trait that associates types to regions, while allowing to specify a owned type that is different to `Self`. Switches reachability logging to send flat containers over dataflow edges. Changes `mz_replay` to accept a closure to produce data.

Related: #27741

Part of #13038

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by frankmcsherry — APPROVED** [review]
Looks good, with some comments about clarity and stumbling blocks I found.

File: src/compute/src/logging/initialize.rs [hunk 1]
```diff
@@ -77,16 +77,25 @@ pub fn initialize<A: Allocate + 'static>(
...
 }
 
+/// Type to specify the region for holding reachability events. Only intended to be interpreted
+/// as [`MzContainerized`].
+type ReachabilityEvent = (
+    OwnedRegionMarker<usize>,
+    OwnedRegionMarker<(usize, usize, bool, Option<Timestamp>, Diff)>,
+);
+pub(super) type ReachabilityRegion =
+    <(Duration, WorkerIdentifier, ReachabilityEvent) as MzContainerized>::Region;
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/compute/src/logging/initialize.rs]: Going with `MzRegionPreference` for now. It's easy to revisit should we want to change it.

File: src/compute/src/logging.rs [hunk 2]
```diff
@@ -35,93 +36,92 @@ use crate::typedefs::RowRowAgent;
...
+
+impl<CB, P, D> PushInto<D> for BatchLogger<CB, P>
+where
+    CB: ContainerBuilder + PushInto<D>,
+    P: EventPusher<Timestamp, CB::Container>,
+{
+    fn push_into(&mut self, item: D) {
+        self.builder.push_into(item);
+        self.extract_and_send();
+    }
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/compute/src/logging.rs]: Moved `extract_and_send` to the `push_into` call sites.

File: src/compute/src/logging/initialize.rs [hunk 3]
```diff
@@ -153,67 +162,65 @@ impl<A: Allocate + 'static> LoggingContext<'_, A> {
...
+        let mut logger =
+            BatchLogger::<CapacityContainerBuilder<_>, _>::new(event_queue.link, self.interval_ms);
         Logger::new(
             self.now,
             self.start_offset,
             self.worker.index(),
             move |time, data| {
-                logger.publish_batch(time, data);
+                logger.push_container(data);
+                logger.flush_through(time);
```
[the remark this answers is not in the document]
Comment (antiguru) [reply src/compute/src/logging/initialize.rs]: Brought back `pusblish_batch` Makes a bunch of things simpler.

**Pull request merged**